src/audio/chain_dma.c

```diff
@@ -37,56 +48,478 @@ static struct comp_dev *chain_dma_create(const struct comp_driver *drv,
 	if (!cd)
 		goto error;
 
-	cd->host_connector_node_id = cdma->primary.r.host_dma_id;
-	cd->link_connector_node_id = cdma->primary.r.link_dma_id;
 	cd->first_data_received = false;
 	cd->under_over_run_notification_sent = false;
+	cd->cs = scs ? 2 : 4;
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+
 	comp_set_drvdata(dev, cd);
 
+	ret = chain_task_init(dev, host_dma_id, link_dma_id, fifo_size);
+	if (ret)
+		goto error;
+
 	return dev;
 
 error:
 	rfree(dev);
 	return NULL;
 }
 
-int chain_dma_start(struct comp_dev *dev)
+int chain_task_start(struct comp_dev *dev)
 {
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	k_spinlock_key_t key;
+	struct comp_driver_list *drivers = comp_drivers_get();
+	int ret;
+
+	comp_info(dev, "chain_task_start(), host_dma_id = 0x%08x", cd->host_connector_node_id.dw);
+	if (cd->chain_task.state == SOF_TASK_STATE_QUEUED)
+		return IPC4_SUCCESS;
+
+	key = k_spin_lock(&drivers->lock);
+	if (!(cd->chain_task.state == SOF_TASK_STATE_COMPLETED
+	    || cd->chain_task.state == SOF_TASK_STATE_INIT))
+		return IPC4_BAD_STATE;
+	cd->under_over_run_notification_sent = false;
+
+	ret = chain_link_start(dev);
+	if (ret)
+		goto error;
+
+	ret = chain_host_start(dev);
+	if (ret)
+		goto error;
+
+	pm_policy_state_lock_get(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
+
+	ret = schedule_task(&cd->chain_task, 0, 0);
+	if (ret)
+		goto error;
+
+	cd->chain_task.state = SOF_TASK_STATE_INIT;
+	k_spin_unlock(&drivers->lock, key);
+
 	return IPC4_SUCCESS;
+
+error:
+	k_spin_unlock(&drivers->lock, key);
+	return IPC4_FAILURE;
 }
 
-int chain_dma_pause(struct comp_dev *dev)
+int chain_task_pause(struct comp_dev *dev)
 {
+	struct chain_dma_data *cd = comp_get_drvdata(dev);
+	struct comp_driver_list *drivers = comp_drivers_get();
+	k_spinlock_key_t key;
+
+	if (cd->chain_task.state == SOF_TASK_STATE_FREE)
+		return IPC4_SUCCESS;
+
+	key = k_spin_lock(&drivers->lock);
+	cd->first_data_received = false;
+
+	if (chain_stop(dev) < 0) {
+		k_spin_unlock(&drivers->lock, key);
+		return IPC4_FAILURE;
+	}
+
+	cd->chain_task.state = SOF_TASK_STATE_COMPLETED;
+	k_spin_unlock(&drivers->lock, key);
+	schedule_task_free(&cd->chain_task);
+	pm_policy_state_lock_put(PM_STATE_RUNTIME_IDLE, PM_ALL_SUBSTATES);
```

Comment:
The `pm_policy_state_lock_get()` is protected by the `drivers->lock` while here it is outside?